daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/SpecifiedLength2.scala

```diff
@@ -460,6 +463,64 @@ class ElementUnusedUnparser(
 
 }
 
+/**
+ * This is basically a simplified version of the ElementUnusedUnparserSuspendableOperation.
+ * For explicit length choices, we do not need to handle length expressions, choiceLength is
+ * simply a non-negative integer.
+ */
+class ChoiceUnusedUnparserSuspendableOperation(
+  override val rd: TermRuntimeData,
+  targetLengthInBits: Long)
+  extends SuspendableOperation {
+
+  /**
+   * determine delta between value length and target length
+   *
+   * and skip that many bits.
+   */
+  override def continuation(ustate: UState) {
+    val e = ustate.currentInfosetNode.asInstanceOf[DIElement]
```

Comment:
I'm wondering if we might run into issues by usign the value length of the current infoset node. For example, say we have a single complex element that is a sequence of two choiceLengthKind="explicit" choices:
```xsd
<xs:element name="foo">
  <xs:complexType>
    <xs:sequence>
      <xs:choice dfdl:choiceLengthKind="explicit" dfdl:choiceLength="X">
        ....
      </xs:choice>
      <xs:choice dfdl:choiceLengthKind="explicit" dfdl:choiceLength="X">
        ....
      </xs:choice>
    </xs:sequence>
  </xs:complexType>
</xs:element>
```
When we unparse the choice unsed stuff, both choices are going to use the length of foo to determine how much space is unused. But we really want the length of the choice content, not the length of foo. Not sure how to handle this, since there is nothing keeping track of the length of the choice content. Do we need some representation in the infoset for a choice where the length of the choice can be hung when choice length is explicit? I think we had this at one point, but maybe got rid of it?

Reply:
Good spot. This is definitely wrong to use the element's length.
The choice parser/unparser have the explicit length in their state or in the ChoiceRuntimeData object. That's the yardstick, not anything about an enclosing element.

Reply:
Also, let's add a test along the lines of Steve's example above to prove this works in that case.

Reply:
I think we have to be careful around where the state is stored. ChoiceParser/Unparser and ChoiceRUntimeData are both shared between different parses/unparses. So putting state there needs to think about thread safety, the state might need to be in a thread local. I suggested the infoset since it is unique per parse and it has all the logic for dealing with suppressions and calculating lengths when there's relative stuff going on.

Maybe instead of adding a ChoiceNode to the infoset, each ComplexNode in the infoset has both a length state for itself **and** an array of length state for any child choices with explicit lengths? Each choice would need to statically know which state in the array is its own, so that might have some challenges. So the below might look something like:
```scala
val v1 = e.choiceValueLength[staticIndex].lengthInBits.longValue
```

Reply:
I've pushed up a test and it does indeed demonstrate the issues with using valueLength instead of the choiceLength.
Also, for some reason this conversation was hidden by default and I missed it the first time around. Wonder if there is a setting to not hid conversations.